HeterogeneousCore/AlpakaCore/python/ProcessAcceleratorAlpaka.py

```diff
@@ -45,26 +49,44 @@ class ProcessAcceleratorAlpaka(cms.ProcessAccelerator):
     ProcessAcceleratorCUDA) define.
     """
     def __init__(self):
-        super(ProcessAcceleratorAlpaka,self).__init__()
+        super(ProcessAcceleratorAlpaka, self).__init__()
         self._backend = None
+
     # User-facing interface
     def setBackend(self, backend):
         self._backend = backend
+
     # Framework-facing interface
     def moduleTypeResolver(self, accelerators):
         return ModuleTypeResolverAlpaka(accelerators, self._backend)
+
     def apply(self, process, accelerators):
+        if not hasattr(process.MessageLogger, "AlpakaService"):
+            process.MessageLogger.AlpakaService = cms.untracked.PSet()
+
         if not hasattr(process, "AlpakaServiceSerialSync"):
             from HeterogeneousCore.AlpakaServices.AlpakaServiceSerialSync_cfi import AlpakaServiceSerialSync
             process.add_(AlpakaServiceSerialSync)
-        if not hasattr(process, "AlpakaServiceCudaAsync"):
-            from HeterogeneousCore.AlpakaServices.AlpakaServiceCudaAsync_cfi import AlpakaServiceCudaAsync
-            process.add_(AlpakaServiceCudaAsync)
+        process.AlpakaServiceSerialSync.enabled = "cpu" in accelerators
 
-        if not hasattr(process.MessageLogger, "AlpakaService"):
-            process.MessageLogger.AlpakaService = cms.untracked.PSet()
+        if os.system(f"scram tool tag cuda-gcc-support TOOLNAME >& /dev/null") == 0:
```

Comment:
@makortel is there a way to reuse the informati0on from `ProcessAcceleratorCUDA` without checking again here ?

Reply:
If `cuda-gcc-support` is available we build the CUDA version of the Alpaka modules.
Won't they try to access the `AlpakaServiceCudaAsync` to determine if they should run or not ?

Reply:
Thinking of possible ways of running Alpaka modules
1. Module type resolver / module variants: if `not "gpu-nvidia" in accelerators", the CUDA backend modules will never be loaded
2. `SwitchProducerCUDA`: if `not "gpu-nvidia" in accelerators`, the CUDA backend modules will be constructed, but their `produce()` would never get called. Use of `edm::Service<AlpakaServiceCudaAsync>` outside `produce()` would have to check that the service handle is valid.
3. Configuration uses explicitly CUDA-backend modules
   i. The configuration loads the `ProcessAccelerator{CUDA,Alpaka}`: if `not "gpu-nvidia" in accelerators` the job would eventually fail anyway, in this case the message would be along "`AlpakaServiceCudaAsync` service is missing"
   ii. The configuration does not use any of the `ProcessAccelerator{CUDA,Alpaka}`: I'd argue the configuration is in so explicit corner of the phase space that it should explicitly load the services as well.

Did I miss something? If not, as far as I can tell, the only missing piece is that all uses of `edm::Service<*AlpakaService*>` would need to be changed to check the validity of the service handle before accessing the handle. (strictly speaking that would not be necessary for functions called from the `produce()` functions, like `chooseDevice()`, but I'd change them anyway)

Reply:
The more I think about this, the more I think that having a dummy `CUDAService` and `ROCmService` that are always disabled if their backend is not available would make things simpler...

Reply:
> The more I think about this, the more I think that having a dummy `CUDAService` and `ROCmService` that are always disabled if their backend is not available would make things simpler...

On a quick thought that approach would have the consequence that every job using `Configuration.StandardSequences.Accelerators_cff` would end up loading the necessary CUDA and ROCm libraries that the `{CUDA,ROCm}Service` depend on, regardless which (if any) the job would actually use.

Reply:
Isn't that already the case now ?

`Configuration.StandardSequences.Accelerators_cff` loads `HeterogeneousCore.CUDACore.ProcessAcceleratorCUDA_cfi` which loads `ProcessAcceleratorCUDA`, and its `apply()` method makes sure that the `CUDAService` is added to the `process`.
Then it sets `process.CUDAService.enabled` to `True` or `False` depending if `gpu-nvidia` is among the `accelerators`, but the service is always loaded.

Same for ROCm (if available).

Reply:
> Isn't that already the case now ?

Currently yes, but this PR (in its current form) moving away from that. It wasn't really an issue with CUDA alone, as any use of `Configuration.StandardSequences.Accelerators_cff` was in conjunction with NVIDIA GPU. Now that we are adding more platforms, the concern of always loading everything could be something to consider.

Reply:
This configuration will load the `CUDAService`:
```python
import FWCore.ParameterSet.Config as cms

process = cms.Process('TEST')

process.load('Configuration.StandardSequences.Accelerators_cff')

process.source = cms.Source('EmptySource')
process.maxEvents.input = 0
```
resulting in
```
%MSG-i CUDAService:  (NoModuleName) 08-Feb-2023 23:57:26 CET pre-events
CUDAService disabled by configuration
%MSG
```
or something like
```
%MSG-i CUDAService:  (NoModuleName) 08-Feb-2023 23:58:01 CET pre-events
CUDA runtime version 11.5, driver version 11.8, NVIDIA driver version 520.61.05
CUDA device 0: NVIDIA A100-PCIE-40GB (sm_80)
%MSG
```

Reply:
> moving away from that

I'll have to think about it some more.

In general I find it easier to separate the two concerns:
  - `BackendService` is _available_, so _e.g._ a module can reference it
  - `BackendService` is _enabled_ or not, depending on the architecture, the result of `backendIsAvailable`, the value of `process.options.accelerators`, _etc.
  - `BackendService` is _enabled_ or not, depending on the architecture, the result of `backendIsAvailable`, the value of `process.options.accelerators`, _etc.